### ISSUE TITLE
More standard support for date/time validation

### ISSUE THREAD

**Issue by bodiam** [json-schema-org/json-schema-vocabularies#30]
First of all, thanks for your amazing efforts on JSON Schemas, it's a wonderful schema, and it works really well for us.

There's only one issue which we encounter on a regular basis, which is date validation. I've been scouring the documentation and the issues, but I haven't seen a great solution so far.

What we'd like to do, is to validate dates. Unfortunately, the formats of these dates are beyond our control, and aren't in any ISO standard. Often they are delivered in formats such as "YYYYMMDD", "DD-MM-YYYY", DDMMYYYYHHMMSS". I guess you get the idea.

Right now, we use regular expressions to validate these dates, or depending on the validator, we use a custom format (not all validators support this however, and it requires extra implmentation efforts), but it's not ideal (think of leap years for example). It's quite hard. So, I was hoping that the current date/date-time formats could be made a little bit more flexible, and something like a pattern could be used?

We had a few things in mind, such as:

```
"processDate": {
    "type": "string",
    "format": "date-time",
    "pattern": "dd-MM-yyyyy"
}
```

**Pro**: It's quite backwards compatible
**Con**: it's using the pattern regex for something else now. But maybe in combination with format, this might be okay?

or

```
"processDate": {
    "type": "string",
    "format": "dd-MM-yyyyy"
}
```

**Pro**: We believe formats should be used for something like this
**Con**: It's using a custom format, not really in line with the current user defined formats

```
"processDate": {
    "type": "string",
    "format": "dd-MM-yyyyy"
   "pattern": "[0-9]{2}-[0-9]{2}-[1-2][0-9]{3}"
}
```
**Pro**: We believe formats should be used for something like this, plus there's a fallback pattern for those validators not understanding the format
**Con**: It's might be slightly unclear to understand which patterns "wins", especially when they are conflicting.

or:

```
"processDate": {
    "type": "date",
    "format": "dd-MM-yyyyy"
}
```

**Pro**: it's clear that these are dates, in a specific format
**Con**: it's a whole new datatype in JSON schema, which might be more complex to implement.

I understand they are all not ideal, but a date format like this should be doable in most programming languages.

Therefor I was wondering what your thoughts were on this, or even if you had alternatives to the above suggestions.

**Comment by karenetheridge** [comment]
> Someone could design a keyword that is an efficient date templating system (based on strftime or whatever).

This would make an ideal $vocabulary extension. For example, a `strftime` keyword, which could either be strictly an annotation (to indicate to the application how to parse the string), or even used for validation:

```
"properties": {
    "start_date": {
        "type": "string",
        "strftime_pattern": "%Y-%m-%dT%H:%M:%S.%N%z",
    }
}
```

**Comment by awwright** [comment]
> Using a simpler regex, such as [0-9]{4}-[0-9]{2}-[0-9]{2} will work for most dates, but it makes using JSON schemas for input validation a poor choice, since it won't say that 2020-02-30 is an invalid date. And a validation which validates only a subset is hardly validation at all.

There's _always_ multiple passes to validation, and JSON Schema is no exception. That date (`2020-02-30`) actually follows the ABNF grammar laid out in RFC 3339, it is "well formed". What makes it invalid is that it does not follow additional post-processing requirements. Among other things, this includes computing the leap year, day-of-month limits from a lookup table, and leap seconds which, as a rule, are not known until less than a year in advance.

I think the best way to do this is to break apart validation into the tools that do it best. When you're parsing a JSON document, use a JSON parser. When you want to verify the formatting is correct, you use a linter. When you want to test code, you use a unit test suite. When you want to validate the _structure_ of JSON, you use a JSON Schema validator. When strings _inside_ that document specify a date, you parse it with a date parsing library.

All of these possibly generate errors that need to be relayed to the user. And this way, JSON Schema doesn't have to know a thing about leap seconds and leap years, we can just leave that task to a tool that specializes in it.

---

Now I understand there's other purposes where we still want to declare the format of something, even though it's not related to structural validation. This is a case where it makes sense to use a custom vocabulary keyword, probably in the [strftime/strptime](https://linux.die.net/man/3/strftime) format. For example, so I can look at a schema, and know _how_ to construct a date, or how to parse it.


Also, for good error reporting, it does make sense to have some hook into the parser. I think if you have a streaming+validating JSON parser (like I'm working on), you can register a callback that parses the date at the moment it's encountered during parsing, and if your date parsing library emits an error, maybe the parser can pass that error through, preserving line and position information relative to the entire JSON document.

**Comment by MarcGodard** [comment]
Sorry to bump this old issue, but I am also always having issues with date time. My request is related, so though to add here.

I use `{ type: 'string', format: 'date-time' }` and the option `coerceTypes: true`. However, when a data object is passed, I get a validation issue. 1) why isn't  coercion being applied? and 2) What would be the best way to write my own. 

Still a little new to this, and so far love it all but date issues.

**Comment by gregsdennis** [comment]
@MarcGodard I'm not sure what `coerceTypes` is.  It's not anything defined by JSON Schema.  Is it something specific to the implementation you're using?  If so, maybe you need to contact them.

**Comment by MarcGodard** [comment]
@gregsdennis yes, sorry I should have mentioned that I am using `Ajv` to coerce the types. But is string and format `date-time` valid? Or is it just objects?

**Comment by gregsdennis** [comment]
What you put is a valid schema, yes, but `format` isn't validated by default, generally (ajv has a history of doing it's own thing, though, so check with them).

**Comment by MarcGodard** [comment]
Ok thanks. Will do
